---
title: "Tirosh et al. distribution estimation"
output: html_notebook
---

```{r, message=FALSE, warning=FALSE, include=FALSE}
install.packages("pscl")
install.packages("MASS")
install.packages("devtools")

require(devtools)
devtools::install_github('hms-dbmi/scde', build_vignettes = FALSE)

```

```{r}
tumor_cells[1:10,1:10]
dim(tumor_cells)
tc.num.test1 <- data.frame(tumor_cells[1:10,5:50])
head(tc.num.test1)
tc.t.test1 <- data.frame(t(tc.num.test1))
data(es.mef.small)
head(es.mef.small)
str(es.mef.small)

colnames(tc.t.test1 ) <- tumor_cells[1:10,1]
rownames(tc.t.test1 ) <- colnames(tumor_cells)[5:50]
cd <- clean.counts(tc.num.test1, min.lib.size=500, min.reads = 1, min.detected = 1)
cd1 <- clean.counts(es.mef.small, min.lib.size=1000, min.reads = 1, min.detected = 1)
str(es.mef.small)
summary(es.mef.small)
dim(es.mef.small)
str(tc.num.test1)
summary(tc.num.test1[,1:10])
dim(tc.num.test1)
#tc.t.test1 <- data.frame(t(tc.num.test1))
summary(tc.t.test1)

tc.num.test1.unlog=data.frame(apply(tc.t.test1,2,function(x) as.integer(exp(x) - 1)))
#as.integer(exp(tc.num.test1[,2]) -1)

summary(tc.num.test1.unlog)
str(tc.num.test1.unlog)

cd2 <- clean.counts(data.frame(as.matrix(tc.num.test1.unlog)), min.lib.size=1, min.reads = 1, min.detected = 1)
cd3 <- clean.counts(data.frame(as.matrix(tc.t.test1 )), min.lib.size=1, min.reads = 1, min.detected = 1)

o.ifm <- scde.error.models(counts = cd2, n.cores = 4, threshold.segmentation = TRUE, save.crossfit.plots = FALSE, save.model.plots = FALSE, verbose = 1,min.size.entries	 =5)
head(o.ifm)
rownames(o.ifm)
o.prior <- scde.expression.prior(models = o.ifm, counts = cd2, length.out = 400, show.plot = TRUE)
#ediff <- scde.expression.difference(o.ifm, cd2, groups=groups, o.prior, n.randomizations  =  100, n.cores  =  4, verbose  =  1, return.posteriors = TRUE)
#groups <- as.factor(c(rep("1", 1), rep("2",9)))
#ediff$joint.posteriors[[1]]

pst <- scde.posteriors(o.ifm, cd2, o.prior, n.cores = 4,return.individual.posteriors =TRUE)
str(pst)
max(o.prior$x)
str(pst)
plot(y = pst[9,], x = seq(0,6, length.out=401), type= "l")



```

```{r}

require(scde)
data(es.mef.small)

# factor determining cell types
sg <- factor(gsub("(MEF|ESC).*", "\\1", colnames(es.mef.small)), levels = c("ESC", "MEF"))
# the group factor should be named accordingly
names(sg) <- colnames(es.mef.small)  
table(sg)
cd <- clean.counts(es.mef.small, min.lib.size=1000, min.reads = 1, min.detected = 1)
o.ifm <- scde.error.models(counts = cd, groups = sg, n.cores = 4, threshold.segmentation = TRUE, save.crossfit.plots = FALSE, save.model.plots = FALSE, verbose = 1)
data(o.ifm)
head(o.ifm)
valid.cells <- o.ifm$corr.a > 0
table(valid.cells)
o.ifm <- o.ifm[valid.cells, ]
o.prior <- scde.expression.prior(models = o.ifm, counts = cd, length.out = 400, show.plot = FALSE)
o.prior$x
# define two groups of cells
groups <- factor(gsub("(MEF|ESC).*", "\\1", rownames(o.ifm)), levels  =  c("ESC", "MEF"))
names(groups) <- row.names(o.ifm)
# run differential expression tests on all genes.
ediff <- scde.expression.difference(o.ifm, cd, o.prior, groups  =  groups, n.randomizations  =  100, n.cores  =  4, verbose  =  1)
scde.browse.diffexp(ediff, o.ifm, cd, o.prior, groups = groups, name = "diffexp1", port = 1299)
scde.test.gene.expression.difference("Tdh", models = o.ifm, counts = cd, prior = o.prior)




```


```{r}
        # show each posterior
        #layout(matrix(c(1:3), 3, 1, byrow = TRUE), heights = c(2, 1, 2), widths = c(1), FALSE)
        #par(mar = c(2.5, 3.5, 2.5, 3.5), mgp = c(1.5, 0.65, 0), cex = 0.9)
        #par(mar = c(2.5, 3.5, 0.5, 3.5), mgp = c(1.5, 0.65, 0), cex = 0.9)
       # x=pst$post
 #split(x, rep(1:ncol(x), each = nrow(x)))

        pp <- exp(do.call(rbind, lapply(pst$post, function(x) x[9,])))

        cols <- rainbow(nrow(pp), s = 0.8)
        plot(c(), c(), xlim = range(o.prior$x), ylim = range(c(0, pp)), xlab = "expression level", ylab = "individual posterior")
        #if(show.individual.posteriors) {
            lapply(seq_len(nrow(pp)), function(i) lines(o.prior$x, pp[i, ], col = rgb(1, 0.5, 0, alpha = 0.25)))
            plot(c())
            plot(o.prior$x, pp[1, ], col = rgb(1, 0.5, 0, alpha = 0.25), type="l")
            
            
       # }
        #legend(x = ifelse(which.max(na.omit(pjpc)) > length(pjpc)/2, "topleft", "topright"), bty = "n", col = cols, legend = rownames(pp), lty = rep(1, nrow(pp)))
        
        # pjpc <- jpl[[1]]$jp
        # par(new = TRUE)
        # jpr <- range(c(0, na.omit(pjpc)))
            par(mar = c(5, 5,3, 3.5))
        pp <- exp(do.call(rbind, lapply(pst$post, function(x) x[9,])))

        cols <- rainbow(nrow(pp), s = 0.8)
        plot(c(), c(), xlim = range(o.prior$x), ylim = range(c(0, pp)), xlab = "expression level", ylab = "individual posterior")
        lapply(seq_len(nrow(pp)), function(i) lines(o.prior$x, pp[i, ], col = rgb(1, 0.5, 0, alpha = 0.25)))
        par(new = TRUE)
        plot(o.prior$x, pst$jp[9,], axes = FALSE, ylab = "", xlab = "", ylim = range(c(0, na.omit(pst$jp[9,]))), type = 'l', col = 1, lty = 1, lwd = 2)
        axis(4, pretty(pst$jp[9,], 5), col = 1)
        mtext("joint posterior", side = 4, outer = FALSE, line = 2)
        
        dim(pst$post[[1]])
        exp(do.call(rbind, lapply(pst$post, function(x) x[9,])))

```

```{r plot_posterior}
########################################################################################################################
###
### it is an annex function modified from 'scde.expression.difference' function of scde package
###
#################
###
### it takes results from scde.posteriors with return.individual.posteriors = TRUE, the line number - for the gene, scde.expression.prior 
###
#############################################################################################################################
plot_posterior <-
  function(res.scde.posteriors, gene.n, res.scde.expression.prior) {
    par(mar = c(5, 5, 3, 3.5))
    pp <-
      exp(do.call(rbind, lapply(res.scde.posteriors$post, function(x)
        x[gene.n, ])))
    
    cols <- rainbow(nrow(pp), s = 0.8)
    plot(
      c(),
      c(),
      xlim = range(res.scde.expression.prior$x),
      ylim = range(c(0, pp)),
      xlab = "expression level",
      ylab = "individual posterior"
    )
    lapply(seq_len(nrow(pp)), function(i)
      lines(res.scde.expression.prior$x, pp[i,], col = rgb(1, 0.5, 0, alpha = 0.25)))
    par(new = TRUE)
    plot(
      res.scde.expression.prior$x,
      res.scde.posteriors$jp[gene.n, ],
      axes = FALSE,
      ylab = "",
      xlab = "",
      ylim = range(c(0, na.omit(
        res.scde.posteriors$jp[gene.n, ]
      ))),
      type = 'l',
      col = 1,
      lty = 1,
      lwd = 2
    )
    axis(4, pretty(res.scde.posteriors$jp[gene.n, ], 5), col = 1)
    mtext("joint posterior",
          side = 4,
          outer = FALSE,
          line = 2)
    
  }
```

```{r}
plot_posterior(pst, 3, o.prior)
```

```{r}
set_sample= seq(1,20, 1)
prob_sample = sample(seq(0,1, length.out=20))
s_res = sample(set_sample, size = 10000, prob = prob_sample, replace = TRUE)
table(s_res)/10000
length(prob_sample)

````

